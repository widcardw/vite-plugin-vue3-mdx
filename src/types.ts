import type { FilterPattern } from 'vite'
import type { PluggableList } from 'unified'
import type { Options as RemarkRehypeOptions } from 'remark-rehype'

interface Options {
  /**
   * Whether to keep jsx.
   *
   * @default true
   */
  jsx?: boolean

  /**
   * Format of the files to be processed.
   *
   * @default 'mdx'
   */
  format?: 'md' | 'mdx' | 'detect'

  /**
   * Whether to compile to a whole program or a function body.
   *
   * @default 'program'
   */
  outputFormat?: 'program' | 'function-body'

  /**
   * Extensions (with `.`) for markdown.
   */
  mdExtentions?: string[]

  /**
   * Extensions (with `.`) for MDX.
   */
  mdxExtensions?: string[]

  /**
   * List of recma (esast, JavaScript) plugins.
   */
  recmaPlugins?: PluggableList

  /**
   * List of remark (mdast, markdown) plugins.
   */
  remarkPlugins?: PluggableList

  /**
   * List of rehype (hast, HTML) plugins.
   */
  rehypePlugins?: PluggableList

  /**
   * Options to pass through to `remark-rehype`.
   */
  remarkRehypeOptions?: RemarkRehypeOptions

  /**
   * Whether to compile to dynamic import expressions.
   *
   * @default false
   */
  useDynamicImport?: boolean

  /**
   * Resolve imports (and `export â€¦ from`, and `import.meta.url`) from this URL
   */
  baseUrl?: string

  /**
   * Whether to add extra information to error messages in generated code
   *
   * @default false
   */
  development?: boolean

  /**
   * Class names for wrapper div
   *
   * @warning vueJsx cannot parse `<MDXLayout><ChildComponents /></MDXLayout>`
   * into `<div class="markdown-body"><ChildCompoents /></div>`. Instead, the
   * `MDXLayout` element should contain a `slot`, which may be suitable for vue.
   *
   * @default false
   */
  wrapperClasses?: false | string | string[]

  SourceMapGenerator?: any

  /**
   * Custom tranformations apply before and after the markdown transformation
   */
  transforms?: {
    before?: (code: string, id: string) => string
    after?: (code: string, id: string) => string
  }

  /**
   * In order to solve the problem of vue router
   *
   * @default true
   */
  addDisplayName?: boolean

  include?: FilterPattern
  exclude?: FilterPattern
}

interface ResolvedOptions extends Required<Options> {
  wrapperClasses: string | false
}

export {
  Options,
  ResolvedOptions,
}
