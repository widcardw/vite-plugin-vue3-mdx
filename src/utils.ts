function joinClasses(c: false | string | string[]) {
  if (typeof c === 'string')
    return c
  else if (Array.isArray(c))
    return c.filter((i?: string) => i).join(' ')
  return false
}

function addDisplayName(raw: string, id: string) {
  return raw.replace('export default MDXContent', `
MDXContent.displayName = '${id}';
export default MDXContent`.trim())
}

function addLayoutWrapperClasses(raw: string, classes: string) {
  return `
export default function Layout({__mdx_children__}) {
  return <div class="${classes}">{__mdx_children__}</div>;
}

${raw}`
}

export {
  joinClasses,
  addDisplayName,
  addLayoutWrapperClasses,
}
