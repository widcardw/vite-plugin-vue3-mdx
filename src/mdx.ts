import type { TransformResult } from 'vite'
import { compileSync } from '@mdx-js/mdx'
import type { ResolvedOptions } from './types'
import { addDisplayName } from './utils'

function createMDXCompiler(options: ResolvedOptions) {
  return (id: string, raw: string): TransformResult => {
    raw = raw.trimStart()

    if (options.transforms.before)
      raw = options.transforms.before(raw, id)

    // if (options.wrapperClasses)
    //   raw = addLayoutWrapperClasses(raw, options.wrapperClasses)

    const code = compileSync(raw, options)

    return {
      code: addDisplayName(String(code), id),
      map: { mappings: '' } as any,
    }
  }
}

export {
  createMDXCompiler,
}
