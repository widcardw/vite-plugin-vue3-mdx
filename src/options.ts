import type { Options, ResolvedOptions } from './types'
import { joinClasses } from './utils'

function resolveOptions(userOptions: Options): ResolvedOptions {
  const defaultOptions: ResolvedOptions = {
    jsx: true,
    format: 'detect',
    outputFormat: 'program',
    mdExtentions: ['.md', '.markdown', '.mdown', '.mkdn', '.mkd', '.mdwn', '.mkdown', '.ron'],
    mdxExtensions: ['.mdx'],
    recmaPlugins: [],
    remarkPlugins: [],
    rehypePlugins: [],
    remarkRehypeOptions: {},
    wrapperClasses: 'markdown-body',
    transforms: {},
    baseUrl: '.',
    development: false,
    SourceMapGenerator: undefined,
    useDynamicImport: false,
    include: null,
    exclude: null,
  }

  const options = { ...defaultOptions, ...userOptions }
  options.wrapperClasses = joinClasses(options.wrapperClasses)

  return options as ResolvedOptions
}

export {
  resolveOptions,
}
