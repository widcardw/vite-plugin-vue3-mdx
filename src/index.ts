import type { Plugin } from 'vite'
import { createFilter } from 'vite'
import { createMDXCompiler } from './mdx'
import { resolveOptions } from './options'
import type { Options } from './types'

function VitePluginVue3Mdx(userOptions: Options = {}): Plugin {
  const options = resolveOptions(userOptions)

  const mdxToJsx = createMDXCompiler(options)

  const filter = createFilter(
    (userOptions.mdExtentions || []).concat(userOptions.mdxExtensions || []) || /\.mdx?$/,
    userOptions.exclude,
  )
  return {
    name: 'vite-plugin-vue3-mdx',
    transform(raw, id) {
      if (!filter(id))
        return

      try {
        return mdxToJsx(id, raw)
      }
      catch (e: any) {
        this.error(e)
      }
    },
    async handleHotUpdate(ctx) {
      if (!filter(ctx.file))
        return

      const defaultRead = ctx.read
      ctx.read = async function () {
        return mdxToJsx(ctx.file, await defaultRead()).code
      }
    },
  }
}

export default VitePluginVue3Mdx
