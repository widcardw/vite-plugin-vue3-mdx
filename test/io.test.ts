import { readFile } from 'fs/promises'
import { mount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'
import { createMDXCompiler } from '../src/mdx'
import { resolveOptions } from '../src/options'
import { saveToExported } from './utils'

describe('read from file', () => {
  const fileName = 'CounterMdx'
  it('mdx file should be parsed', async () => {
    const mdx = await readFile(`test/files/${fileName}.mdx`, 'utf-8')

    const compiler = createMDXCompiler(resolveOptions({}))
    const jsxCode = (await compiler(fileName, mdx)).code
    await saveToExported(jsxCode, fileName)

    const component = (await import(`./__exported__/${fileName}`)).default
    const wrapper = mount(component)
    expect(wrapper.html()).toMatchInlineSnapshot(`
      "<h1>Hi, this is a counter</h1>



      <p>The counter is 0. Now <button>Click me</button> to increase.</p>"
    `)

    await wrapper.get('button').trigger('click')
    expect(wrapper.text()).toMatchInlineSnapshot(`
      "Hi, this is a counter


      The counter is 1. Now Click me to increase."
    `)
  })
})
