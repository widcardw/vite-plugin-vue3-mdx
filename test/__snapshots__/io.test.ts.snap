// Vitest Snapshot v1

exports[`read from file > mdx file should be parsed 1`] = `
"/*@jsxRuntime automatic @jsxImportSource react*/
import {ref} from 'vue';
export const counter = ref(0);
export const Button = () => <button onClick={() => counter.value++}>Click me</button>;
function _createMdxContent(props) {
  const _components = Object.assign({
    h1: \\"h1\\",
    p: \\"p\\"
  }, props.components);
  return <><_components.h1>{\\"Hi, this is a counter\\"}</_components.h1>{\\"\\\\n\\"}{\\"\\\\n\\"}{\\"\\\\n\\"}<_components.p>{\\"The counter is \\"}{counter.value}{\\". Now \\"}<Button />{\\" to increase.\\"}</_components.p></>;
}
function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = props.components || ({});
  return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
}
MDXContent.displayName = 'CounterMdx';
export default MDXContent;
"
`;
